src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -510,9 +503,51 @@ private bool IsFirstTokenOnLine(TokenData tokenData1, TokenData tokenData2)
 
         private int GetTokenIndexInStream(SyntaxToken token)
         {
-            if (_tokenToIndexMap.TryGetValue(token, out var value))
+            var tokenIndex = _tokens.BinarySearch(token, TokenOrderComparer.Instance);
+            if (tokenIndex < 0)
+            {
+                return -1;
+            }
+
+            // Source characters cannot be assigned to multiple tokens. If the token has non-zero width, it will be an
+            // exact match for at most one token.
+            if (!token.FullSpan.IsEmpty)
             {
-                return value;
+                return _tokens[tokenIndex] == token ? tokenIndex : -1;
+            }
+
+            // Multiple tokens can have empty spans. The binary search operation will return one of them; look forward
+            // and then backward to locate the desired token within the set of one or more zero-width tokens located at
+            // the same position.
```

Comment:
If error tokens aren't included, then this code wouldn't get hit. That seems fine; I wanted to make sure the new implementation followed the old implementation as closely as we could.

Reply:
@heejaechang Would know best here.  however, the core design of roslyn is that you can truly think of there being a real "stream of tokens" where each token is non-zero length.  These are the real tokens effectively, modulo some small exceptions, produced by teh lexer.  These are really the things we think about having whitespace between for formatting.  this token stream round-trips with your text 1:1.  The empty-tokens are really a side effect of the parser needing to fill up the tree with non-default tokens in error cases.  

But, as a true 'stream of tokens' (or 'TokenStream' as this seems to want to represent), we ideally should consider missing-tokens as never appearing. (Ideally...).

Reply:
all tokens including length 0 token should be in the stream. _tokens regardless whether it is 0 length token (can't remember which token, but there was legal 0 length token in Roslyn, compiler team would know) or missing token since trivia is attached to token including missing token.

the dictionary was just reverse map to make token -> index faster for _tokens. so if binary search actually make it faster due to GC, then +1

Reply:
just saw the numbers. so, binary search actually make it slower. 
if we print out tokens that uses the binary search, will same token called multiple time? or always different token? if same token called multiple time, is locality high? or low?

if same token get called multiple time with high locality, we could add simple MRU cache of say 100 tokens and that would reduce overhead a bit?

Reply:
> just saw the numbers. so, binary search actually make it slower

There was a small performance hit in the test scenario, but the locality of the affected code suggests one of two things is likely in practice:

1. Inlining from NGEN could eliminate the observed overhead and closing the performance gap
2. Rewriting the binary search to avoid indirection through `IComparer<SyntaxToken>.CompareTo` could eliminate the observed overhead

Reply:
The test scenario involved formatting around 8500 files (10 iterations of ~850 files). The files in PerfView are relatively large as well (no single class per file limitation and many large files that were generated but not marked as generated code). I can get exact numbers if you want.

Reply:
> all tokens including length 0 token should be in the stream. _tokens regardless whether it is 0 length token (can't remember which token, but there was legal 0 length token in Roslyn, compiler team would know) or missing token since trivia is attached to token including missing token.

Could you explain why that is?  I thought the formatter operated on pairs of adjacent tokens, calling into the formatting rules to decide what to do with them.  Basically to decide "what whitespace change should i make between these two tokens?".  I'm really curious why it would matter if there was an invisible token there.

It would be good to potentially just remove checks for that, and see if any tests break.  Or just throw if we get an empty token here and see what breaks.  We coudl then track this up higher to see what higher level is depending on being able to pass through empty tokens.

Reply:
because I designed that way? it find token to attach trivia based on token in the stream and info It get back from formatting rule which is based on 2 token (previous and current) which include 0 length tokens.

Reply:
➡️ Whether or not 0-length tokens are kept in this long term, I believe there is sufficient evidence to say the search algorithm should at least handle the case.

Reply:
> because I designed that way?

Right.  I get that :)  I'm asking: for what case does that matter?

> it find token to attach trivia based on token in the stream and info It get back from formatting rule which is based on 2 token (previous and current) which include 0 length tokens.

What is an example of a formatting rule that would apply between 0 length tokens?  I'm trying to figure it out.  All the user facing ones would only apply between non-0-length tokens.

--

Note: i get that this is how it's impleemnted *now*, i'm asking the question of: does it need to be implemented this way, or could it be greatly simplified without any negative end user impact?

Reply:
> I believe there is sufficient evidence to say the search algorithm should at least handle the case.

Can you clarify this.  What evidence are you referring to?

> Whether or not 0-length tokens are kept in this long term,

I don't understand the logic here.  If 0 length tokens are not kept int he long term, then why would the search algorithm need to handle that case?

I understand *completely* why we might need to keep this in the short term if we actually have rules that depend on this.  But my question was: can we determine if there are such rules.  And, if not, can we make this simpler in the future?

Reply:
rule provider shouldn't care whether token is missing or whether it is dealing with broken code. rule provider should return rule based on how it wants code to look like in normal case.

the engine will decide what it will do if rules it got back is related to broken (missing token).

for specific case where someone want to deal with very specific error case, they might return different rule based on tokens it got, but generally that is special case for like code generation and etc.

in normal case, rule provider should not care what code it is called upon, it should just return what space/lines it wants between 2 tokens they are adjacent.

Reply:
> But my question was: can we determine if there are such rules. And, if not, can we make this simpler in the future?

ha ha, that's very basic design decision I made on the engine. and that's basic assumption on implementation. that token stream will return every tokens parse tree will have. what you are saying is basically re-writing whole thing since it is breaking the core assumption. 

that's like why Roslyn is immutable, change it to mutable will greatly improve allocations.

but anyway, sure, if you want to rewrite and other people are okay with it. why not?

Reply:
> what you are saying is basically re-writing whole thing since it is breaking the core assumption.

Can you give an example of code that depends on that assumption?

> but anyway, sure, if you want to rewrite and other people are okay with it. why not?

I'm not asking about a rewriting.  I'm simply asking if we could potentially ignore missing tokens, and what impact that would be on existing rules.  Could you give an example of a rule that needs to be able to run on missing tokens in order to work properly?

Reply:
can't remember example.

make the change that you want to verify assumption and run formatter and see what breaks. remove zero length token from _token or any field that contains token data and see what breaks.